app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -73,49 +74,71 @@ def batch_callbacks(batch)
     end
   end
 
-  def delete_associations
-    destroyable_associations.each do |association|
-      ReflectionDestroyer.new(object, association, caller_worker_hierarchy).destroy_later
+  def destroy_and_delete_associations
+    Array(object.background_deletion).each do |association, options|
+      ReflectionDestroyer.new(object, association, options, caller_worker_hierarchy).destroy_later if destroyable_association?(association)
     end
   end
 
-  private
-
-  def destroy_method
-    @destroy_method.presence || 'destroy'
+  def destroyable_association?(_association)
+    true
   end
 
+  private
+
   def called_from_provider_hierarchy?
     return unless (tenant_id = object.tenant_id)
     caller_worker_hierarchy.include?("Hierarchy-Account-#{tenant_id}")
   end
 
-  def destroyable_associations
-    object.class.reflect_on_all_associations.select { |reflection| destroyable_association?(reflection) }
-  end
-
-  def destroyable_association?(reflection)
-    %i[destroy delete_all].include?(reflection.options[:dependent])
-  end
-
   def callback_options
     { 'object_global_id' => object.to_global_id, 'caller_worker_hierarchy' => caller_worker_hierarchy }
   end
 
   class ReflectionDestroyer
 
-    def initialize(main_object, reflection, caller_worker_hierarchy)
+    class FakeReflection
+      attr_reader :name, :options
+
+      def initialize(name, options)
+        @name = name
+        @options = options
+      end
+
+      def many?
+        return options[:has_many] if options.keys.include?(:has_many)
+
+        plural?
+      end
+
+      def class_name
+        options[:class_name].presence || name.to_s.singularize.classify
```

Comment:
I hope that it isn't a problem, but just in case... `options[:class_name]` is a String and `name.to_s.singularize.classify` is a class 🤔 perhaps it is safer to do:
`options[:class_name].presence || name.to_s.singularize.classify.name`

Reply:
In fact I think that if it is a class and not a string, it will not do the conditional well in here:
https://github.com/3scale/porta/pull/1372/files#diff-ff63dfaf8bfa4a850361d41a988711a1R165

Reply:
`classsify` does not return class, but rather a class name: https://apidock.com/rails/ActiveSupport/Inflector/classify